### DISCUSSION TITLE
cell background and event placing on footer (Orig.Title:Question about the design)

### DISCUSSION THREAD

**Discussion by m4nino** [MMRIZE/MMM-CalendarExt3#109] [Q&A] [answered]
Is it possible to do this? 

The first one, the M have the end time 9 and the N have the start time 10, but IRL it does'nt look like this. I've made the picture in photoshop. 

The second one, can I change something to always put this event in the bottom of the calendar? 

![2023-11-07-123241_1920x1080_scrot_000](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/f0ec96d1-2c85-4310-9588-17d3932613c5)

**Comment by eouia** [comment]
> The first one, the M have the end time 9 and the N have the start time 10, but IRL it does'nt look like this. I've made the picture in photoshop.


1. It looks somewhat ideal but makes non-sense. How does it look with 20 min-period events?
For this kind of purpose, another timeline view is needed instead. [`MMM-CalendarExt3Timeline`](https://github.com/MMRIZE/MMM-CalendarExt3Timeline) might be.

<img width="1246" alt="image" src="https://github.com/MMRIZE/MMM-CalendarExt3/assets/1720610/208849c4-54fe-4d12-b30a-5299b68dc117">

> The second one, can I change something to always put this event in the bottom of the calendar?

Very challenging question. To place the top position, you can use `eventSorter`, 
```js
eventSorter: (a, b) => {
	if (a.calendarName === 'ManU') return -1
},
```
For example, By this custom function, if the events from `ManU` calendar, that events will be placed on top (if possible) regardless of it's event time

But placing the last position with empty lines? Well, this module is not designed so. This module cannot do that by itself.

3. But... With another module, it can be forced to happen.

![image](https://github.com/MMRIZE/MMM-CalendarExt3/assets/1720610/a3ab32ff-2636-440e-8e91-67f4cd5213cf)

You need [`MMM-ModuleMonkeyPatch`](https://github.com/MMRIZE/MMM-ModuleMonkeyPatch)

```js
{
	module: "MMM-ModuleMonkeyPatch",
	config: {
		patches: [
			{
				module: "MMM-CalendarExt3",
				method: "getDom",
				patch: function (original, args) {
					let dom = original(args)
					let target = Array.from(dom.querySelectorAll('.event')) || []
					target.forEach((e) => {
						if (e.dataset.calendarName === 'ManU') { // If calendar name is "ManU", put that event to cellFooter area.
							const dt = new Date(+e.dataset.startDate)
							const date = '.date_' + dt.getDate()
							const month = '.month_' + (dt.getMonth() + 1)
							const container = dom.querySelector(`.cell${date + month} .cellFooter`)
							if (container) container.appendChild(e)
						}
					})
					return dom
				}
			}
		]
	}
},
```

And some CSS tweek. (in `css/custom.css`)
```css
.CX3 {
  --cellfooterheight: 25px;
}

.CX3 .cell {
  overflow: hidden;
}

.CX3 .cellFooter {
  overflow: hidden;
}
```


This example doesn't regard error exceptions. Just show you it is possible.

**Reply by m4nino** [reply]
I will try the module. For the first one, would it be possible to paint the background in the cell instead of show it as an event? I have seen someone make an background in one cell for halloween or some other holiday. Maybe that could be usefull? I will attach another image from photoshop (from my cellphone). Have a look at the pink and blue background. 

![IMG_5564 (1)](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/ee7d219c-1bd7-4577-805a-c2abee91c593)

**Reply by eouia** [reply]
I’m out of desk now so cannot give you details at this moment. I’ll check this tomorrow.
By the way, you won’t give up the idea of intersection in a single day(22nd day). Is it really essential for your purpose?

**Reply by m4nino** [reply]
Haha, I will be more than happy for a solutions even without the intersection, but that would have been totally on point. 

I have my children half time, change with their father on wednesdays. Therefore Im thinking about an nice way to add the schedule into the calendar. It’s not static though, have to be possible to easy change color for day if we change the schedule one week.

**Reply by eouia** [reply]
Interesting.
- there would be only 2 conditions at any moment, A or B, not else.
- only one intersection possible in a day, not more than once.
Right?

**Reply by m4nino** [reply]
Yes, that’s right!

**Comment by eouia** [comment, marked as answer]
![image](https://github.com/MMRIZE/MMM-CalendarExt3/assets/1720610/2b28eda1-4aa1-40d6-b173-740faa1c3d48)

## Conditions
- Only consider event `TurnA` and `TurnB`, (no TurnC or other else)
- Distinguish events by only obvious event titles. 
  - Identifiers are `TurnA` and `TurnB`. e.g. `BlahBlahTurnABlah` will be regarded as `TurnA`. But `turn A` will not be recognized.
  - `TurnA and TurnB` will also be problem, so keep the rule.
 - Each event should be long enough to over night. 
  
 ## Configuration
 ```js
 // In config of module CX3
 manipulateDateCell: (cell, events) => {
	if (cell.dataset.events < 1) return
	const cellDate = new Date(+cell.dataset.date)
	const date = cellDate.getDate()
	events.forEach((event) => {
		const keyMap = {
			'a' : 'TurnA', // 'TurnA' is the part of event title to target, 'a' is the class name to assign
			'b' : 'TurnB',
		}
		if (!Object.values(keyMap).some((name) => { return event.title.includes(name) })) return
		for (const [ classKey, searchTitle ] of Object.entries(keyMap)) {
			if (event.title.includes(searchTitle)) {
				cell.classList.add(classKey)
				if (date === (new Date(+event.startDate)).getDate()) cell.classList.add(classKey + '-start')
				if (date === (new Date(+event.endDate)).getDate()) cell.classList.add(classKey + '-end')
			}
		}
	})
}
 ```
To adjust to your case, modify `TurnA` and `TurnB` in `keyMap` parts of the source. You have to choose a title carefully that is unique to separate from other normal events.(e.g. `@Father`, `@Mother`) 
 
 ## CSS
 ```css
 .CX3 {
  --color-a: rgb(255 200 200 / 50%);
  --color-b: rgb(200 200 255 / 50%);
}

.CX3 .cell.a {
  background-color: var(--color-a);
}

.CX3 .cell.b {
  background-color: var(--color-b);
}

.CX3 .cell.a-start.b-end {
  background: linear-gradient(110deg, var(--color-b) 40%, var(--color-a) 40%);
}

.CX3 .cell.b-start.a-end {
  background: linear-gradient(110deg, var(--color-a) 40%, var(--color-b) 40%);
}
 ```

**Reply by m4nino** [reply]
![11](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/0ce9c888-d1c0-4188-b343-d1452cf41a38)

```
  { 
             module: "MMM-CalendarExt3",
             position: "bottom_bar",
             title: "month",
             config: 
             {
             weekIndex: 0,
             locale: 'sv-SE',
             maxEventLines: 10,
             eventHeight: '22px',
             firstDayOfWeek: 1,
             calendardfSet: ['a', 'TurnB'],
             useSymbol: false,
             useWeather: false,
  },
               manipulateDateCell: (cell, events) => {
   if (cell.dataset.events < 1) return
   const cellDate = new Date(+cell.dataset.date)
   const date = cellDate.getDate()
   events.forEach((event) => {
   	const keyMap = {
   		'a' : 'TurnA', // 'TurnA' is the part of event title to target, 'a' is the class name to assign
   		'b' : 'TurnB',
   	}
   	if (!Object.values(keyMap).some((name) => { return event.title.includes(name) })) return
   	for (const [ classKey, searchTitle ] of Object.entries(keyMap)) {
   		if (event.title.includes(searchTitle)) {
   			cell.classList.add(classKey)
   			if (date === (new Date(+event.startDate)).getDate()) cell.classList.add(classKey + '-start')
   			if (date === (new Date(+event.endDate)).getDate()) cell.classList.add(classKey + '-end')
   		}
   	}
   })
}
},  
		   {
       module: "calendar",
			config: {
				broadcastPastEvents: true, // <= IMPORTANT to see past events
				updateInterval: 1000 * 60,
				calendars: [
					{	
						url: "https://calendar.google.com/calendar/ical/575c4316d34a1ea4cf872815860654abee4302ad60f29f975a2a1cad08f54a31%40group.calendar.google.com/private-7c5cba90519aa7ccd0d8af6b1365d79e/basic.ics",
						name: 'a', // <= RECOMMEND to assign name
						color: "#facfef", // <= RECOMMEND to assign color
					},
				    {	
						url: "https://calendar.google.com/calendar/ical/b792c19776e5439bad216922421a7361b7ad450a1847bf4b638cf2a14b95f03c%40group.calendar.google.com/private-d00cb02ffc759cf1bcf08b4c3e702afe/basic.ics",
						name: 'TurnB', // <= RECOMMEND to assign name
						color: "#c8defa", // <= RECOMMEND to assign color
					},
					
	```
   .CX3 {
  /* you CAN modify these values; but SHOULD NOT to remove */
  background-color: rgba(255, 255, 255, 1);
  border-radius:20px;
  padding: 14px;
  --color-a: rgb(255 200 200 / 50%);
  --color-b: rgb(200 200 255 / 50%);
  --celllinecolor: #718187;   
  --cellbgcolor: #000000;     
  --cellheaderheight: 20px;
  --cellfooterheight: 15px;
  --defaultcolor: #444345;
  --eventheight: calc(var(--fontsize) + 4px);
  --totalheight: calc(var(--eventheight) * var(--maxeventlines));
  font-size: 25px;
  color: var(--defaultcolor);
  line-height: calc(var(--eventheight));
  }  

.CX3 .cell.a {
 background-color: var(--color-a);
}

.CX3 .cell.b {
 background-color: var(--color-b);
}

.CX3 .cell.a-start.b-end {
 background: linear-gradient(110deg, var(--color-b) 40%, var(--color-a) 40%);
}

.CX3 .cell.b-start.a-end {
 background: linear-gradient(110deg, var(--color-a) 40%, var(--color-b) 40%);
}

**Reply by eouia** [reply]
?

**Reply by eouia** [reply]
![image](https://github.com/MMRIZE/MMM-CalendarExt3/assets/1720610/8f9c6911-8618-46d6-91fe-de7f908e69ed)
Does not it look like your purpose?

**Reply by eouia** [reply]
Ah, Here.
```js
 config: 
             {
             weekIndex: 0,
             locale: 'sv-SE',
             maxEventLines: 10,
             eventHeight: '22px',
             firstDayOfWeek: 1,
             calendardfSet: ['a', 'TurnB'],
             useSymbol: false,
             useWeather: false,
  }, // <=== Here.
               manipulateDateCell: (cell, events) => {
               ..
```
`manipulateDateCell`... should be included in the config of module. Your bracket makes this property out.

**Reply by eouia** [reply]
and there is `calendarSet` mistypo in your config also. (Not affected though.)

**Reply by m4nino** [reply]
First of all! Thank you, for letting me know the mistypo. I have been wondering why the calendarSet did'nt work for me, now it does! 

And then, the intersection thing and the colored background is amazing. Look at it. That's a perfect view. You made it!!

But the monkeymodule, I do something wrong. The purple event is gone when I add it with copy paste css and config from you. If I change the overflow to visible, I can see it, but they are drowning. They are supposed to be on the 14th and 15th november. I have tried to change the cellfooterheight with no difference. Can you see something strange? Maybe it's collide with something else in my css? I have changed alot in the css overall so maybe this is'nt going to work for me. 

![2023-11-10-215847_1920x1080_scrot_000](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/5b74752c-7853-48f1-af66-41bce9d56e4b)
![2023-11-10-223135_1920x1080_scrot](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/9e36a252-0c36-4b49-8c40-eefedbc2fbb8)

### Config 

``` 
{
	module: "MMM-ModuleMonkeyPatch",
	config: {
		patches: [
			{
				module: "MMM-CalendarExt3",
				method: "getDom",
				patch: function (original, args) {
					let dom = original(args)
					let target = Array.from(dom.querySelectorAll('.event')) || []
					target.forEach((e) => {
						if (e.dataset.calendarName === "Eskola") { // If calendar name is "Eskola", put that event to cellFooter area.
							const dt = new Date(+e.dataset.startDate)
							const date = '.date_' + dt.getDate()
							const month = '.month_' + (dt.getMonth() + 1)
							const container = dom.querySelector(`.cell${date + month} .cellFooter`)
							if (container) container.appendChild(e)
						}
					})
					return dom
				}
			}
		]
	}
}, 
```
### CSS

```

.CX3 {
  /* you CAN modify these values; but SHOULD NOT to remove */
  background-color: rgba(255, 255, 255, 1);
  border-radius:20px;
  padding: 14px;
  --color-a: rgb(255 118 169 / 10%);
  --color-b: rgb(189 227 255 / 25%);
  --celllinecolor: #718187;   
  --cellbgcolor: #000000;     
  --cellheaderheight: 25px;
  --cellfooterheight: 25px;
  --defaultcolor: #444345;
  --eventheight: calc(var(--fontsize) + 4px);
  --totalheight: calc(var(--eventheight) * var(--maxeventlines));
  font-size: 25px;
  color: var(--defaultcolor);
  line-height: calc(var(--eventheight));
  }  

.CX3 .cell {
  overflow: hidden;
  border: 1.8px solid var(--celllinecolor);
  background-color: var(--cellbgcolor);
  height: 250px;
  margin: 1.8%;
  margin-bottom: 0%;
  background-color: rgb(255 255 255 / 87%);
  border-radius: 22px;  
}  

.CX3 .cellFooter {
  overflow: hidden;
}

**Reply by eouia** [reply]
Without full configuration and whole CSS, I cannot tell you anything. But, what I found is; 
The height of cell is CALCULATED by many factors like this.
```css
.CX3 .cell {
...
  height: calc(
    var(--cellheaderheight) + var(--cellfooterheight) +
      calc(var(--eventheight) * var(--maxeventlines))
  );
...
}
```
And --eventheight is calculated from --fontsize, and others goes on like this. And some times these values would be changed dynamically, so auto-calculation is essential.

But you override this as `250px` fixed value. So it will not match with other calculated values. 
How to solve this;

- Calculate all related height by your hand, and assign proper values.
- Or, just adjust other value instead of `250px`  to the cell to match with module's calculation. Many shoot & aim trials would be needed.  
- Or, just adjust fontsize, eventheight, maxeventlines, cellheader, cellfooter ... to match with your ideal design. And adjust also minor properties for that purpose.

Sorry, this module plays CSS too heavily, so even a small change could make a big butterfly effect.

**Comment by m4nino** [comment]
I did got the monkeymodule right after a while. I did remove the CSS and copy paste back step by step. But then, I made a mistake again and the view went  almost like this. AND I LOVE IT. Thank you for giving me this. The first idea was to have another calendar in the monkeymodule, but this went better than that! I really appreciate it. Simple, nice view. 

![2023-11-12-001711_1920x1080_scrot](https://github.com/MMRIZE/MMM-CalendarExt3/assets/145551766/096d4793-2630-4859-8548-b75b5521e981)
